### ISSUE TITLE
Issues running on WSL2

### ISSUE THREAD

**Issue by Poimen** [sakaicontrib/docker-sakai-builder#15]
Thanks for the scripts and dockerfiles, it's been extremely useful getting up and running.

We wanted to use the default release and placed it into `/work/tomcat/deploy`. I couldn't find early 19.x releases, so used 21.1.

When it runs up, I get the error:
```
INFO: OpenSSL successfully initialized [OpenSSL 1.1.1k  25 Mar 2021]
[1.955s][info][gc] GC(0) Pause Young (Normal) (G1 Evacuation Pause) 56M->16M(1026M) 16.017ms
[1.976s][info][gc] GC(1) Pause Young (Normal) (G1 Evacuation Pause) 59M->15M(1026M) 7.614ms
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.apache.ignite.internal.util.GridUnsafe$2 (file:/usr/src/app/deploy/lib/ignite-core-2.9.1.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of org.apache.ignite.internal.util.GridUnsafe$2
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[2.192s][info][gc] GC(2) Pause Young (Normal) (G1 Evacuation Pause) 85M->16M(1026M) 2.311ms
```

The later:
```
[20:32:42] To start Console Management & Monitoring run ignitevisorcmd.{sh|bat}
06-Dec-2021 20:32:42.782 INFO [main] org.apache.ignite.internal.IgniteKernal%localhost.info To start Console Management & Monitoring run ignitevisorcmd.{sh|bat}
[20:32:42] 
[20:32:42] Ignite node started OK (id=d566f6d8, instance name=localhost)
06-Dec-2021 20:32:42.783 INFO [main] org.apache.ignite.internal.IgniteKernal%localhost.info 

>>> +----------------------------------------------------------------------+
>>> Ignite ver. 2.9.1#20201203-sha1:adcce517ce542fdaca954bce399e7a1940630403
>>> +----------------------------------------------------------------------+
>>> OS name: Linux 5.4.72-microsoft-standard-WSL2 amd64
>>> CPU(s): 12
>>> Heap: 2.0GB
>>> VM name: 1@dbd17fb43e10
>>> Ignite instance name: localhost
>>> Local node [ID=D566F6D8-B0CF-47C9-9677-F524B66F0B36, order=1, clientMode=false]
>>> Local node addresses: [dbd17fb43e10/127.0.0.1, /172.17.0.3]
>>> Local ports: TCP:10800 TCP:11211 TCP:64629 TCP:64639 
```

And the final error:
```
06-Dec-2021 20:32:44.518 INFO [main] org.sakaiproject.springframework.orm.hibernate.impl.AdditionalHibernateMappingsImpl.processAdditionalUnit Add mapping resource [org/sakaiproject/sitestats/impl/hbm/UserStat.hbm.xml]
06-Dec-2021 20:32:44.669 INFO [main] org.sakaiproject.ignite.IgniteEntityManagerFactoryBean.setIgnite Ignite instance name [localhost] configured as hibernate cache provider
06-Dec-2021 20:32:45.318 WARN [main] org.sakaiproject.util.SakaiApplicationContext.refresh Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in file [/usr/src/app/deploy/components/sakai-kernel-impl/WEB-INF/db-components.xml]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException
06-Dec-2021 20:32:45.329 INFO [main] org.sakaiproject.log.impl.Log4jConfigurationManager.destroy destroy()
06-Dec-2021 20:32:45.333 INFO [main] org.apache.ignite.internal.processors.rest.protocols.tcp.GridTcpRestProtocol.info Command protocol successfully stopped: TCP binary
06-Dec-2021 20:32:45.355 ERROR [main] org.apache.ignite.internal.IgniteKernal%localhost.error Failed to stop component (ignoring): GridProcessorAdapter []
java.lang.UnsupportedOperationException
	at org.jsr166.ConcurrentLinkedHashMap.clear(ConcurrentLinkedHashMap.java:1542)
	at org.apache.ignite.internal.processors.job.GridJobProcessor.stop(GridJobProcessor.java:376)
	at org.apache.ignite.internal.IgniteKernal.stop0(IgniteKernal.java:2577)
...
```

I have wsl configured:
```
[wsl2]
size=8G
processors=12
```
so there should be enough memory and processors for it...

Not sure if this is an error with the images, release or something in configuration?

Any help would be appreciated.

**Comment by jonespm** [comment]
Hi!

I run on WSL2 (Windows 11) but I haven't tried 21 in awhile, as this is mostly setup to build the master branch. This won't work with anything earlier as Java 11 and MySQL 8 (MariaDB) weren't supported until Sakai 21. It's possible older versions of this might work for those since I for sure was using it initially in 2018 for Sakai 19.

Have you tried it with the more recent 21.2 tag or the 21.x branch? Or even at least the master to see if that starts up?

I could check my wsl2 config but I'm expecting I set it for 4G and 4 processors, so that shouldn't be the issue.

**Comment by Poimen** [comment]
Thanks - I've been running around today trying a bunch of different options.

I managed to build `19.x` branch and still no luck, but moved back to the `21.1` binary release. I haven't tried building the `21.x` branch today.

I'm not sure what you mean by:
> This won't work with anything earlier as Java 11 and MySQL 8 (MariaDB) weren't supported until Sakai 21`

Do you mean the docker commands in the repo?

**Comment by jonespm** [comment]
The error line seems to be

`06-Dec-2021 20:32:45.318 WARN [main] org.sakaiproject.util.SakaiApplicationContext.refresh Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in file [/usr/src/app/deploy/components/sakai-kernel-impl/WEB-INF/db-components.xml]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException`

Ahh, I think the issue is that Sakai 21 only supported Java 8 and earlier. Sakai 22 was Java 11. This script was updated to Java 11 and support for Sakai 22 so only will **probably** run Sakai 22. (Which is the 22.x branch, or master which is 23)

You'd probably have to go back to an older revision before JDK 11 https://github.com/sakaicontrib/docker-sakai-builder/commit/a77c0d8b6e269549a6e5f26c09fa194a149b1430

**Comment by jonespm** [comment]
I was remembering this when I saw this on the manual install guide. https://confluence.sakaiproject.org/pages/viewpage.action?pageId=122618517

So you'll need to use JDK 8 and might even have to go back to MySQL 5.7 for Sakai 19-21. Sakai 22 supports JDK 11. I don't know if MySQL 5.8 works as I've been using Maria for awhile now.

**Comment by jonespm** [comment]
Sorry, I mistyped. I'd though that JDK 11 worked for MySQL 21 but it doesn't. So these JDK images used in this script won't work for Sakai 21 unless you revert them back the java version to 8.

**Comment by Poimen** [comment]
Interesting - I've tried so many options today, I think I may be getting myself confused on different versions...

I've built my own building maven container - as readme hinted at:
```
FROM maven:3.6.1-jdk-8-slim
```

With docker containers, it's pretty simple to move 🛩️ 

Ok, so to confirm - Sakai 21 (want to stick to release versions) should be JDK 8 on MySql 5 - which should be fine for that build container? It could be that in the runtime it was using JDK 11?

**Comment by Poimen** [comment]
Just checked my logs:
```
INFO: Server version name:   Apache Tomcat/9.0.55
INFO: Server built:          Nov 10 2021 08:26:45 UTC
INFO: Java Home:             /usr/local/openjdk-11
```

So that'll have to move the JDK-8 as I'm assuming `21.x` is the latest version for production use...

The QA Nightly for `21.x`:
```
07-Dec-2021 16:40:58.174 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Server version name:   Apache Tomcat/9.0.43
07-Dec-2021 16:40:58.176 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Server built:          Jan 28 2021 20:25:45 UTC
07-Dec-2021 16:40:58.176 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Server version number: 9.0.43.0
07-Dec-2021 16:40:58.176 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log OS Name:               Linux
07-Dec-2021 16:40:58.176 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log OS Version:            5.4.0-1030-aws
07-Dec-2021 16:40:58.176 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Architecture:          amd64
07-Dec-2021 16:40:58.177 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Java Home:             /usr/lib/jvm/java-8-openjdk-amd64/jre
07-Dec-2021 16:40:58.177 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log JVM Version:           1.8.0_292-8u292-b10-0ubuntu1~20.04-b10
```

So JDK-8 on that.

**Comment by jonespm** [comment]
Right, if you're running it off the commands in this shell script, line 62 here uses tomcat with JDK 11. I think you can just change that to `tomcat:9-jdk8-openjdk`

https://github.com/sakaicontrib/docker-sakai-builder/blob/cb7e280b73ed6b79692090c685524fff92a96e83/sakai-dock.sh#L62

Might also have to remove the `JDK11_OPTS` from the `JAVA_OPTS` line.

**Comment by Poimen** [comment]
After some option twiddling I got it to work 😄 

For any future reference, this was my process - 
1. Built the `mavenbuild` docker image
2. Updated `MAVEN_IMAGE="sakai:build"` and removed the `--pull always` on the maven build step
3. Checked out the `tags/21.1` release for Sakai
4. Build the code base using the instructions
5. Dropped `mysql...` connector into the `deploy` lib directory
6. Updated the `mariadb:10` image to `mysql:5.7.16` (mariadb kept crashing when Sakai tried to create indexes) (also remembering to set `MYSQL_ROOT_PASSWORD` env variable)
7.  Updated the Tomcat image to `tomcat:9-jdk8-openjdk` with removing `JDK11_OPTS` and `JDK11_GC` from the `JAVA_OPTS` as well as adding `-Djava.net.preferIPv4Stack=true` (there was also a strange `"` that I removed 🤷‍♂️ )
8. Changed the `JPDA_ADDRESS` to `"JPDA_ADDRESS=8000"` for Tomcat 9 on JDK 8

After all this I got to the portal of Sakai and seems to be working 🥇 🎉 

Thanks for the support 👍
